### PR TITLE
Allow multi-flavours projects

### PR DESCRIPTION
**Pull request by kopiczko**
Towards https://github.com/giantswarm/giantswarm/issues/15611

Tested with:
- config-controller: generated from my local build: https://github.com/giantswarm/config-controller/pull/83
- config-controller: 0.2.3-alpha1 release: https://github.com/giantswarm/config-controller/pull/84
- Created release https://github.com/giantswarm/config-controller/releases/tag/v0.2.3-alpha1
- Updated collection: https://github.com/giantswarm/aws-app-collection/commit/3166a5ebc64e6f512215df039361d4a4548ab290

### CODE REVIEW

File: pkg/gen/flavour.go [hunk 1]
```diff
@@ -39,7 +40,92 @@ func (f Flavour) String() string {
 	return string(f)
 }
 
-func IsValidFlavour(s string) bool {
-	_, err := NewFlavour(s)
-	return err == nil
+type FlavourSlice []Flavour
```
Comment (kopiczko) [inline pkg/gen/flavour.go:43]: This beauty here is adjusted copy of pflag.intSlice

File: pkg/gen/language.go [hunk 2]
```diff
@@ -39,3 +40,30 @@ func IsValidLanguage(s string) bool {
 	_, err := NewLanguage(s)
 	return err == nil
 }
+
+type LanguageFlagValue Language
```
Comment (kopiczko) [inline pkg/gen/language.go:44]: This beauty here is adjusted copy of pflag.intValue

File: cmd/gen/makefile/flag.go [hunk 3]
```diff
@@ -16,21 +16,29 @@ const (
 )
 
 type flag struct {
-	Flavour  string
-	Language string
+	Flavours gen.FlavourSlice
+	Language gen.Language
```
Comment (kopiczko) [inline cmd/gen/makefile/flag.go:20]: Using custom values simplified the code inside runners nicely. Validation is now straightforward. But the important bit is that `Flavour  string ` is replaced with `Flavours gen.FlavourSlice`.

File: cmd/gen/makefile/flag.go [hunk 4]
```diff
@@ -16,21 +16,29 @@ const (
...
-	cmd.Flags().StringVarP(&f.Language, flagLanguage, "l", "", fmt.Sprintf(`The programming language of project that you want to generate the Makefile for. Possible values: <%s>`, strings.Join(gen.AllLanguages(), "|")))
+	cmd.Flags().VarP(gen.NewFlavourSliceFlagValue(&f.Flavours, []gen.Flavour{}), flagFlavour, "f", fmt.Sprintf(`List of types of project that you want to generate the Makefile for. Possible values: <%s>`, strings.Join(gen.AllFlavours(), "|")))
+	cmd.Flags().VarP(gen.NewLanguageFlagValue(&f.Language, gen.Language("")), flagLanguage, "l", fmt.Sprintf(`The programming language of project that you want to generate the Makefile for. Possible values: <%s>`, strings.Join(gen.AllLanguages(), "|")))
 }
 
 func (f *flag) Validate() error {
-	if !gen.IsValidFlavour(f.Flavour) {
-		return microerror.Maskf(invalidFlagError, "--%s must be one of <%s>", flagFlavour, strings.Join(gen.AllFlavours(), "|"))
+	if len(f.Flavours) == 0 {
+		return microerror.Maskf(invalidFlagError, "--%s must not be empty", flagFlavour)
```
Comment (mig4) [inline cmd/gen/makefile/flag.go:30]: Would there be a way to still show allowed values? E.g. `must be one or more of ...`

Comment (kopiczko) [reply cmd/gen/makefile/flag.go:30]: Feel free to suggest something. `gen.AllFlavours()` is still there. I thought it's enough to have them printed in `--help`.

**Comment by kopiczko** [comment]
This is tested. See the updated description.

**Review by MarcelMue — APPROVED** [review]
Works for me.

File: cmd/gen/makefile/flag.go (hunk 4 above)
Comment (mig4) [reply cmd/gen/makefile/flag.go:30]: I find the best errors are those that tell me how to fix them 😉 

```suggestion
		return microerror.Maskf(invalidFlagError, "--%s must be one or more of: %s", flagFlavour, strings.Join(gen.AllFlavours(), ", ")
```

File: cmd/gen/makefile/flag.go [hunk 5]
```diff
@@ -16,21 +16,29 @@ const (
...
 	}
-	if !gen.IsValidLanguage(f.Language) {
-		return microerror.Maskf(invalidFlagError, "--%s must be one of <%s>", flagLanguage, strings.Join(gen.AllLanguages(), "|"))
+	if len(f.Language) == 0 {
+		return microerror.Maskf(invalidFlagError, "--%s must not be empty", flagLanguage)
```
Comment (mig4) [inline cmd/gen/makefile/flag.go:33]:
```suggestion
		return microerror.Maskf(invalidFlagError, "--%s must be one of: %s", flagLanguage, strings.Join(gen.AllLanguages(), ", "))
```

File: cmd/gen/workflows/flag.go [hunk 6]
```diff
@@ -17,17 +17,17 @@ const (
...
-	cmd.Flags().StringVarP(&f.Flavour, flagFlavour, "f", "", fmt.Sprintf(`The type of project that you want to generate the workflows for. Possible values: <%s>`, strings.Join(gen.AllFlavours(), "|")))
+	cmd.Flags().VarP(gen.NewFlavourSliceFlagValue(&f.Flavours, gen.FlavourSlice{}), flagFlavour, "f", fmt.Sprintf(`The type of project that you want to generate the workflows for. Possible values: <%s>`, strings.Join(gen.AllFlavours(), "|")))
 	cmd.Flags().BoolVar(&f.CheckSecrets, flagCheckSecrets, true, "If true, also generate a secret-scanning workflow. Possible values: true (default), false.")
 }
 
 func (f *flag) Validate() error {
-	if !gen.IsValidFlavour(f.Flavour) {
-		return microerror.Maskf(invalidFlagError, "--%s must be one of <%s>", flagFlavour, strings.Join(gen.AllFlavours(), "|"))
+	if len(f.Flavours) == 0 {
+		return microerror.Maskf(invalidFlagError, "--%s must not be empty", flagFlavour)
```
Comment (mig4) [inline cmd/gen/workflows/flag.go:30]:
```suggestion
		return microerror.Maskf(invalidFlagError, "--%s must be one of: %s", flagFlavour, strings.Join(gen.AllFlavours(), ", "))
```

**Review by mig4 — CHANGES_REQUESTED** [review]

**Review by kubasobon — APPROVED** [review]
LGTM

**Review by mig4 — APPROVED** [review]
Thank you ❤️ 🚀

**Pull request merged**